### PR TITLE
Add new gnuplot capabilities (surface, contour, contour3d, scatter3d, wireframe, heatmap, ...)

### PR DESCRIPTION
**Pull request by t-bltg**
Add new gnuplot capabilities:

```julia
using Plots; gaston()

wframe() = begin
  x = [0, 1, 2, 3]
  y = [0, 1, 2]
  z = [10 10 10 10; 10 5 1 0; 10 10 10 10]
  wireframe(x, y, z, title="3D: Valley of the Gnu from gnuplot manual", legend=:none)
  png("wframe")
end

scatter3() = begin
  x = 0:.1:6π
  scatter3d(x, cos.(x), sin.(x))
  png("scatter3")
end

sombrero() = begin
  x = y = -15:.4:15
  surface(x, y, (x, y) -> @.(sin(sqrt(x^2 + y^2)) / sqrt(x^2 + y^2)), legend=:none, color=:viridis)
  png("sombrero")
end

hmap() = begin
  x = y = -5:.1:5
  heatmap(x, y, (x, y) ->cos.(x / 2) .* sin.(y / 2), legend=:none)
  png("hmap")
end

cnt() = begin
  x = y = -5:.1:5
  contour(x, y, (x, y) -> 5cos.(x / 2) .* sin.(y / 2), legend=:none, aspect_ratio=:equal)
  png("cnt")
end

cnt3d() = begin
  x = y = -5:.1:5
  contour3d(x, y, (x, y) -> 5cos.(x / 2) .* sin.(y / 2), legend=:none)
  png("cnt")
end

wframe()
scatter3()
sombrero()
hmap()
cnt()
cnt3d()
```

wframe
scatter3
sombrero
hmap
cnt
cnt3d

### CODE REVIEW

**Comment by isentropic** [comment]
Do you know how does it work out with grid layout?

**Comment by t-bltg** [comment]
> Do you know how does it work out with grid layout?

What do you mean ?

**Comment by isentropic** [comment]
Have you tested if multiplots (layout, multiple subplots, 1 figure), work normal when 3d plots are considered?

**Comment by t-bltg** [comment]
> Have you tested if multiplots (layout, multiple subplots, 1 figure), work normal when 3d plots are considered?

Fails:
```julia
multiplot3d() = begin
  x = 0:.1:6π
  p1 = scatter3d(x, cos.(x), sin.(x), legend=:none)
  x = y = -15:.4:15
  p2 = surface(x, y, (x, y) -> @.(sin(sqrt(x^2 + y^2)) / sqrt(x^2 + y^2)), legend=:none)
  t = .01:.01:10π
  p3 = plot(t, cos.(t), legend=:none)
  plot((p1, p2, p3)..., layout=@layout([_ _ °; _ ° _; ° _ _]))
  png("multiplot3d")
end
```

, let me see what I can do here.

**Comment by t-bltg** [comment]
`contour` and `contour3d` are now working !

File: src/backends/gaston.jl [hunk 1]
```diff
@@ -129,45 +149,91 @@ function gaston_add_series(plt::Plot{GastonBackend}, series::Series)
...
+    series[:linewidth],
+)
+
+gaston_mk_ms_mc(series::Series) = (
+    gaston_marker(series[:markershape]),
+    series[:markersize] * GASTON_MARKER_SCALING,
+    gaston_color(series[:markercolor], series[:markeralpha]),
+)
+
+gaston_palette(gradient) = begin
```
Comment (isentropic) [inline src/backends/gaston.jl:183]: Here lets have full function decorators with explicit return

**Review by isentropic — CHANGES_REQUESTED** [review]
Overall looks very good to me. Thanks a lot for your contribution. Also, let me know what you think in general about this backend? I personally like gnuplot for what it is (was), are the benefits of gnuplot (terminals, interactiveness, oldschool feel, performance)   well reflected?

**Comment by isentropic** [comment]
What could be our selling point for http://docs.juliaplots.org/latest/backends ?
We should also mention that given the age of gnuplot some of the plots functionality can't be achieved at all

**Comment by t-bltg** [comment]
> What could be our selling point for http://docs.juliaplots.org/latest/backends ?
> We should also mention that given the age of gnuplot some of the plots functionality can't be achieved at all

Let's discuss that https://github.com/JuliaPlots/PlotDocs.jl/pull/250 ?

**Comment by t-bltg** [comment]
> Overall looks very good to me. Thanks a lot for your contribution. Also, let me know what you think in general about this backend? I personally like gnuplot for what it is (was), are the benefits of gnuplot (terminals, interactiveness, oldschool feel, performance) well reflected?

Well I've used gnuplot a few years ago, mostly because my colleagues were using it for automated tests in CFD codes. It is fast, robust, but most of the users are a bit afraid by the command line. It reaches publication quality, whereas you have to tweak pyplot a lot to reach something acceptable ...

**Comment by t-bltg** [comment]
@isentropic, are you ok with the latest changes ?

**Review by isentropic — APPROVED** [review]
Looks good to me. Feel free to merge when you feel like it.

**Pull request merged**